### PR TITLE
Fix #2521: CommandButton: add option to disable button on Ajax requests

### LINKED ISSUE #2521: Prevent double click on UICommand components
This topic is quite recurrent, would it be possible to integrate a solution in PF?

https://stackoverflow.com/questions/10756426/how-to-do-double-click-prevention-in-jsf-2
http://ovaraksin.blogspot.fr/2011/05/prevent-multiple-clicks-in-jsf.html

Components:
- CommandButton
- CommandLink
- MenuItem
- SplitButton
- HotKey

### LINKED ISSUE #8213: CommandButton: add option to disable button on Ajax requests
Add option to disable the button on Ajax requests triggered by that button.

### PR DESCRIPTION
**Pull request by jepsar**
Fix #2521

### CODE REVIEW

**Comment by Rapster** [comment]
Could we do this for every UICommand in PF?

**Comment by jepsar** [comment]
If we would add `disable` to `SplitButton`, we can add it there. `MenuButton` has `deactivate` but it requires a menu item to disable. We might want to disable the entire component.

I don't think disabling other menus will make sense.

`CommandLink` has no widget.

**Comment by Rapster** [comment]
As long it's possible to invoke an action, it makes sense. Would it be a lot of work to make commandLink a widget? But wait a sec, what does this fix then if not that issue https://www.primefaces.org/showcase-ext/sections/renderers/commandButtonSingleClick.jsf ?

**Comment by jepsar** [comment]
We can have a look at adding a widget for `CommandLink` to disable it. #8218

I don't understand your last question. The renderer in PFE can be deprecated and removed if this PR is accepted.

**Comment by Rapster** [comment]
Ok cool I just wanted to make sure of that

**Comment by tandraschko** [comment]
i would like to see the "getSourceIs" in our ajax.js as util and maybe reimplement as "isXhrSource(...)"

**Comment by jepsar** [comment]
> i would like to see the "getSourceIs" in our ajax.js as util and maybe reimplement as "isXhrSource(...)"

Okay. I'll also refactor the block UI JavaScript.

**Comment by jepsar** [comment]
I'll create a new PR for block UI.. it was causing to much noise on this PR

File: primefaces/src/main/resources/META-INF/resources/primefaces/core/core.ajax.js [hunk 1]
```diff
@@ -148,6 +148,32 @@ if (!PrimeFaces.ajax) {
...
+            /**
+             * Checks whether the component ID from the provided configuration equals the source ID from the provided
+             * settings.
+             *
+             * @param {PrimeFaces.PartialWidgetCfg<TCfg>} cfg of the component to check for being the source.
+             * @param {JQuery.AjaxSettings} settings containing source ID.
+             * @returns {boolean} `true` if the component ID from the provided configuration equals the source ID from
+             * the provided settings.
+             */
+            isXhrSource: function(cfg, settings) {
```
Comment (tandraschko) [inline primefaces/src/main/resources/META-INF/resources/primefaces/core/core.ajax.js]: i would either pass the id or the whole widget

Comment (tandraschko) [reply primefaces/src/main/resources/META-INF/resources/primefaces/core/core.ajax.js]: not sure ;D i dont see any reason to pass a wrapper object from the middle of the stack, isntead of the whole widget or only the id

**Pull request merged**